### PR TITLE
Fixes #10259 - Templates are now fully proxied

### PR DESCRIPTION
**Pull request by lzap**
It was ignoring some parameters (MAC for bootdisk) as well as headers
(HTTP_X_RHN_PROVISIONING_*).

Also removed incorrect TFTP unit test that was failing on all systems which
do not have set TFTP root to `/var/lib/tftpboot`.

Changes by Lukas Zapletal

Replaces https://github.com/theforeman/smart-proxy/pull/287

### CODE REVIEW

**Comment by lzap** [comment]
@ohadlevy this implements what we talked about - full proxying. Also fixed RHN Anaconda headers which were totally ignored!

**Comment by lzap** [comment]
Fixed Ruby 1.8.7 syntax error.

**Comment by lzap** [comment]
And a Rubocop warning.

File: test/tftp/tftp_test.rb [hunk 1]
```diff
@@ -11,10 +11,6 @@ def test_should_have_a_logger
     assert_respond_to @tftp, :logger
   end
 
-  def test_path_to_tftp_directory_without_tftproot_setting
```
Comment (domcleal) [inline test/tftp/tftp_test.rb:14]: This looks unrelated.

**Comment by ohadlevy** [comment]
@witlessbird can you review please? :+1: from my side.

**Comment by ohadlevy** [comment]
@lzap is there a PR in foreman side to take usage of this functionality ?

File: modules/templates/templates_api.rb [hunk 2]
```diff
@@ -14,8 +14,8 @@ class Proxy::TemplatesApi < Sinatra::Base
   end
 
   get "/:kind" do |kind|
-    log_halt(500, "Failed to retrieve #{kind} template for #{params[:token]}: ") do
-      Proxy::Templates::Handler.get_template(kind, params[:token], params[:static])
+    log_halt(500, "Failed to retrieve #{kind} template for #{params.inspect}: ") do
+      Proxy::Templates::Handler.get_template(kind, params, headers)
```
Comment (dmitri-d) [inline modules/templates/templates_api.rb]: `Sinatra::Helpers#headers` is for setting response headers, and returns current response headers when called without parameters: http://www.rubydoc.info/gems/sinatra/Sinatra/Helpers#headers-instance_method. Rack stores request headers in `request.env` hash, under keys starting with `HTTP_` (http://rack.rubyforge.org/doc/SPEC.html, lookup "HTTP_ Variables").

File: modules/templates/handler.rb [hunk 3]
```diff
@@ -7,23 +7,24 @@ module Proxy::Templates
...
 
-    def get_template(kind, token, static = false)
-      opts = { :token => token,
-               :url   => Proxy::Templates::Plugin.settings.template_url
-      }
-      opts[:static] = static if static
+    def get_template(kind, params, headers)
+      opts = params.merge(:url => Proxy::Templates::Plugin.settings.template_url)
       request = request_factory.create_get("/unattended/#{kind}", opts)
+      headers.each do |k, v|
```
Comment (witlessbird) [inline modules/templates/handler.rb]: This modifies response headers, not request headers of `request`.

File: modules/templates/handler.rb [hunk 4]
```diff
@@ -7,23 +7,24 @@ module Proxy::Templates
...
+        request.add_field(k, v)
```
Comment (dmitri-d) [inline modules/templates/handler.rb]: Probably worth modifying `ForemanRequestFactory` to accept headers as a parameter to `create_` methods (https://github.com/theforeman/smart-proxy/blob/develop/lib/proxy/request.rb#L16).

File: modules/templates/handler.rb [hunk 5]
```diff
@@ -7,23 +7,24 @@ module Proxy::Templates
...
+      end
```
Comment (dmitri-d) [inline modules/templates/handler.rb]: Also, I think it would be nice to add `X-Forwarded-For` header here, since we are acting as an http proxy.

File: modules/templates/templates_api.rb (hunk 2 above)
Comment (lzap) [inline modules/templates/templates_api.rb]: Not sure what you mean, I want to verbosely copy all the headers from request to response, so `HTTP_` prefix is unwanted and the helper method `headers` is not avaible in the Handler instance.

**Comment by lzap** [comment]
@ohadlevy https://github.com/theforeman/foreman_bootdisk/pull/9

File: modules/templates/handler.rb (hunk 4 above)
Comment (lzap) [inline modules/templates/handler.rb]: Sure will do.

File: modules/templates/handler.rb (hunk 5 above)
Comment (lzap) [inline modules/templates/handler.rb]: Ok good point. I have amended one additional change because it was not working properly. Will amend the above two points later today.

File: modules/templates/templates_api.rb (hunk 2 above)
Comment (dmitri-d) [inline modules/templates/templates_api.rb]: at https://github.com/theforeman/smart-proxy/pull/300#discussion-diff-34666024R18 you are calling `Sinatra::Helpers#headers` to populate headers of the request to Foreman. This method doesn't return incoming request headers, you'll get something like: 
`{"Content-Type"=>nil}`
vs. actual request headers:
`{"HTTP_USER_AGENT"=>"curl/7.32.0", "HTTP_HOST"=>"localhost:8000", "HTTP_ACCEPT"=>"*/*", "HTTP_VERSION"=>"HTTP/1.1"}`

**Comment by lzap** [comment]
Amended, I am passing all three objects in the Handler instance:

```
      Proxy::Templates::Handler.get_template(kind, request, params, headers)
```

I think this is the sanest method.

File: modules/templates/handler.rb [hunk 6]
```diff
@@ -7,23 +7,29 @@ module Proxy::Templates
...
-      opts = { :token => token,
-               :url   => Proxy::Templates::Plugin.settings.template_url
-      }
-      opts[:static] = static if static
-      request = request_factory.create_get("/unattended/#{kind}", opts)
-      logger.debug "retrieving a template from %s/%s" % [Proxy::Templates::Plugin.settings.template_url, request.path]
-      res = send_request(request)
+    def get_template(kind, request, params, headers)
+      url = Proxy::Templates::Plugin.settings.template_url || raise("Setting template_url not set!")
+      proxy_ip = URI.parse(url).host rescue "127.0.0.1"
```
Comment (dmitri-d) [inline modules/templates/handler.rb]: Instead of this and the line above, set a default value for `template_url` setting in https://github.com/theforeman/smart-proxy/blob/develop/modules/templates/templates_plugin.rb

Comment (lzap) [inline modules/templates/handler.rb]: Actually this must be always true, otherwise the plugin disables itself. I will remove the extra code.

**Comment by lzap** [comment]
Fixed tests and all the comments. Added extra test file for ForemanRequest class.

File: modules/templates/handler.rb [hunk 7]
```diff
@@ -7,23 +7,28 @@ module Proxy::Templates
...
-      logger.debug "retrieving a template from %s/%s" % [Proxy::Templates::Plugin.settings.template_url, request.path]
-      res = send_request(request)
+    def get_template(kind, request, params)
+      url = Proxy::Templates::Plugin.settings.template_url
+      proxy_ip = URI.parse(url).host
+      opts = params.clone.merge(:url => url)
+      opts.delete("kind")
+      opts.delete("splat")
+      opts.delete("captures")
+      proxy_headers = Hash[request.env.select{|k,v| k =~ /^HTTP_/ && k !~ /^HTTP_(VERSION|HOST)$/}.map{|k,v| [k[5..-1],v]}].merge("X-Forwarded-For" => "#{request.ip}, #{proxy_ip}")
```
Comment (dmitri-d) [inline modules/templates/handler.rb]: This will break on 1.8.7 (Hash#select returns an array there).

File: modules/templates/handler.rb [hunk 8]
```diff
@@ -7,23 +7,28 @@ module Proxy::Templates
   class Handler < ::Proxy::HttpRequest::ForemanRequest
```
Comment (dmitri-d) [inline modules/templates/handler.rb]: Could you rename `Handler` to something like `ProxyTemplateRequest`?

File: modules/templates/handler.rb [hunk 9]
```diff
@@ -7,23 +7,28 @@ module Proxy::Templates
...
 
       # You get a 201 from the 'built' URL
-      raise "Error retrieving #{kind} for #{token} from #{uri.host}: #{res.class}" unless ["200", "201"].include?(res.code)
+      raise "Error retrieving #{kind} for #{opts.inspect} from #{uri.host}: #{res.class}" unless ["200", "201"].include?(res.code)
+      Proxy::Log.logger.info "Template: request for #{kind} using #{opts.inspect} at #{uri.host}"
       res.body
     end
 
-    def self.get_template kind, token, static = false
+    def self.get_template(kind, request, params)
```
Comment (dmitri-d) [inline modules/templates/handler.rb]: I would prefer passing ip and env hash directly, it's more clear and easier to test...

File: modules/templates/handler.rb [hunk 10]
```diff
@@ -7,23 +7,28 @@ module Proxy::Templates
...
-               :url   => Proxy::Templates::Plugin.settings.template_url
-      }
-      opts[:static] = static if static
-      request = request_factory.create_get("/unattended/#{kind}", opts)
-      logger.debug "retrieving a template from %s/%s" % [Proxy::Templates::Plugin.settings.template_url, request.path]
-      res = send_request(request)
+    def get_template(kind, request, params)
+      url = Proxy::Templates::Plugin.settings.template_url
+      proxy_ip = URI.parse(url).host
+      opts = params.clone.merge(:url => url)
```
Comment (dmitri-d) [inline modules/templates/handler.rb]: I don't think you need to clone `params` here, `ForemanRequestFactory` isn't going to modify the object...

Comment (lzap) [inline modules/templates/handler.rb]: Factory does not modify it, but `get_template` method does. It is removing some keys from it below.

Comment (dmitri-d) [inline modules/templates/handler.rb]: Are these the modifications below? Do you know where these keys are coming from?

File: modules/templates/handler.rb (hunk 7 above)
Comment (lzap) [inline modules/templates/handler.rb]: I don't understand, this is a clean code and works fine.

File: modules/templates/handler.rb (hunk 8 above)
Comment (lzap) [inline modules/templates/handler.rb]: Sure, fixed.

File: modules/templates/handler.rb (hunk 9 above)
Comment (lzap) [inline modules/templates/handler.rb]: Hmmm I am not so sure about this one. Objects should be passed instead of it's data. I prefer nicer code to nicer tests.

File: modules/templates/handler.rb (hunk 10 above)
Comment (lzap) [inline modules/templates/handler.rb]: Yes these are generated by Sinatra. Since I want this to be generic, I want to copy all the parameters except these. Not sure if there is more clever method than this one.

File: modules/templates/template_proxy_request.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    include Proxy::Log
+
+    def get_template(kind, request, params)
+      url = Proxy::Templates::Plugin.settings.template_url
+      proxy_ip = URI.parse(url).host
+      opts = params.clone.merge(:url => url)
+      opts.delete("kind")
+      opts.delete("splat")
+      opts.delete("captures")
+      proxy_headers = Hash[request.env.select{|k,v| k =~ /^HTTP_/ && k !~ /^HTTP_(VERSION|HOST)$/}.map{|k,v| [k[5..-1],v]}].merge("X-Forwarded-For" => "#{request.ip}, #{proxy_ip}")
```
Comment (dmitri-d) [inline modules/templates/template_proxy_request.rb]: Didn't realise ruby does a splat under the hood when assigning an array to multiple vars, so this works under 1.8.7. None of the tests cover the filter bit though, which would be useful. Probably worth extracting this line into a method too.

File: test/templates/template_proxy_request_test.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+require 'webmock/test_unit'
+
+
+class TemplateProxyRequestTest < Test::Unit::TestCase
+  def setup
+    @foreman_url = 'https://foreman.example.com'
+    Proxy::SETTINGS.stubs(:foreman_url).returns(@foreman_url)
+    @template_url = 'http://proxy.lan:8443'
+    Proxy::Templates::Plugin.settings.stubs(:template_url).returns(@template_url)
+    @request = mock('request')
```
Comment (dmitri-d) [inline test/templates/template_proxy_request_test.rb]: This mocking wouldn't be necessary if you passed environment and ip parameters directly into `TemplateProxyRequest.get_template` method. Not only would this make it clear what the method operates on, but would also make tests less fragile. Don't use mocks/stubs unless absolutely necessary.

File: modules/templates/template_proxy_request.rb (hunk 11 above)
Comment (lzap) [inline modules/templates/template_proxy_request.rb]: Adding tests, actually passing just `env` hash is better in this case.

File: test/templates/template_proxy_request_test.rb (hunk 12 above)
Comment (lzap) [inline test/templates/template_proxy_request_test.rb]: Yeah, fixed.

**Comment by lzap** [comment]
Amended.

File: test/templates/template_proxy_request_test.rb [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    @foreman_url = 'https://foreman.example.com'
+    Proxy::SETTINGS.stubs(:foreman_url).returns(@foreman_url)
+    @template_url = 'http://proxy.lan:8443'
+    Proxy::Templates::Plugin.settings.stubs(:template_url).returns(@template_url)
+    @request_env = {
+      'REMOTE_ADDR' => '1.2.3.4'
+    }
+  end
+
+  def text_header_extraction_empty
```
Comment (dmitri-d) [inline test/templates/template_proxy_request_test.rb]: `test` is misspelled here and in two other tests below, which means they'll be skipped...

**Comment by dmitri-d** [comment]
rubocop isn't happy either. Looks good other than that.

**Comment by lzap** [comment]
Thanks for review!

**Comment by lzap** [comment]
Fixed failed tests.

**Comment by lzap** [comment]
@witlessbird any chance of merging that in before your PTO? Thanks!

**Comment by dmitri-d** [comment]
@lzap: will test this in a bit.

**Comment by dmitri-d** [comment]
merged in at 38a0935. Thanks, @lzap!
